src/mame/machine/jaleco_ms32_sysctrl.cpp

```diff
@@ -0,0 +1,337 @@
+	  causing SFX overloads by using Spitfire ship with 30 Hz autofire 
+	  and shooting at point blank range over walls/enemies. 
+	  This has been fixed in v1.1
+
+*******************************************************************************/
+
+#include "emu.h"
+#include "jaleco_ms32_sysctrl.h"
+
+//*****************************************************************************
+//  GLOBAL VARIABLES
+//*****************************************************************************
+
+// device type definition
+DEFINE_DEVICE_TYPE(JALECO_MS32_SYSCTRL, jaleco_ms32_sysctrl_device, "jaleco_ms32_sysctrl", "Jaleco MS32 System Control Unit")
+
+
+//*****************************************************************************
+//  LIVE DEVICE
+//*****************************************************************************
+
+//-------------------------------------------------
+//  jaleco_ms32_sysctrl_device - constructor
+//-------------------------------------------------
+
+jaleco_ms32_sysctrl_device::jaleco_ms32_sysctrl_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, JALECO_MS32_SYSCTRL, tag, owner, clock)
+	, m_screen(*this, finder_base::DUMMY_TAG)
+	, m_flip_screen_cb(*this)
+	, m_vblank_cb(*this)
+	, m_field_cb(*this)
+	, m_prg_timer_cb(*this)
+	, m_sound_reset_cb(*this)
+{
+}
+
+void jaleco_ms32_sysctrl_device::amap(address_map& map)
+{
+//	0xba0000 in 68k, 0xfce00000 in MS32 mapped at lower 16-bits mask
+	map(0x00, 0x01).w(FUNC(jaleco_ms32_sysctrl_device::control_w));
+	map(0x02, 0x03).w(FUNC(jaleco_ms32_sysctrl_device::hblank_w));
+	map(0x04, 0x05).w(FUNC(jaleco_ms32_sysctrl_device::hdisplay_w));
+	map(0x06, 0x07).w(FUNC(jaleco_ms32_sysctrl_device::hbp_w));
+	map(0x08, 0x09).w(FUNC(jaleco_ms32_sysctrl_device::hfp_w));
+	map(0x0a, 0x0b).w(FUNC(jaleco_ms32_sysctrl_device::vblank_w));
+	map(0x0c, 0x0d).w(FUNC(jaleco_ms32_sysctrl_device::vdisplay_w));
+	map(0x0e, 0x0f).w(FUNC(jaleco_ms32_sysctrl_device::vbp_w));
+	map(0x10, 0x11).w(FUNC(jaleco_ms32_sysctrl_device::vfp_w));
+	map(0x18, 0x19).w(FUNC(jaleco_ms32_sysctrl_device::timer_interval_w));
+	map(0x1a, 0x1b).w(FUNC(jaleco_ms32_sysctrl_device::timer_ack_w));
+	map(0x1c, 0x1d).w(FUNC(jaleco_ms32_sysctrl_device::sound_reset_w));
+	map(0x1e, 0x1f).w(FUNC(jaleco_ms32_sysctrl_device::irq_ack_w));
+//	map(0x24, 0x27).w // sound comms bidirectional acks?
+	map(0x28, 0x29).nopw(); // watchdog on MS32
+	map(0x2c, 0x2d).w(FUNC(jaleco_ms32_sysctrl_device::field_ack_w));
+	map(0x2e, 0x2f).w(FUNC(jaleco_ms32_sysctrl_device::vblank_ack_w));
+}
+
+//-------------------------------------------------
+//  device_add_mconfig - device-specific machine
+//  configuration addiitons
+//-------------------------------------------------
+
+void jaleco_ms32_sysctrl_device::device_add_mconfig(machine_config &config)
+{
+	TIMER(config, "scantimer").configure_scanline(FUNC(jaleco_ms32_sysctrl_device::scanline_cb), m_screen, 0, 1); 
+
+	// TODO: watchdog
+}
```

Comment:
You really shouldn’t be doing things that depend on your own object finders being finalised at this point.

Reply:
I'm curious about the possible alternative here, that isn't just using two extra `emu_timer`s which makes me puke plus won't work in eventually extending it.

Reply:
Well, `timer_device` is just a less efficient way to use `emu_timer`.  When your `timer_device` fires, it has to go through all these steps:
* Delegate to `device_timer` trampoline (indirect call, increases L1 cache and TLB load)
* vtable dispatch to `timer_device::device_timer` (another indirect call)
* Switch on `m_type` (conditional branch hurts pipeline efficiency)
* Check `m_first_time` (another conditional branch)
* Check whether `m_callback` is set (yet another conditional branch)
* Call through delegate to your timer handler (yet another indirect call)

They’re absolutely horrible for anything that’s called multiple times per frame.  You can avoid most of that stuff by using a delegate-style `emu_timer` instead.  Even using an `emu_timer` that goes through your own `device_timer` skips quite a few of the performance-sapping steps.

If you really insist on using a `timer_device` you can set its parameters from your `set_screen`, but remember you have to use `subdevice` to find it because your own object finders are cleared out between when `device_add_mconfig` returns and object resolution time (necessary because of potential to remove/replace devices from outside).

Reply:
Hmm, interesting.
I tend to use `configure_scanline` because it's easier to eventually change (and we don't yet know the specifics about the field irq). I guess I may just create an `emu_timer` given that our `screen_device` cannot generate any callback that isn't the vblank interrupt.